### PR TITLE
Updated to support API changes for Mambu 3.12

### PR DESCRIPTION
**Pull request by mdanilkis**
1. Updated pom.xml and the source code to use 3.12 Mambu model jar
2. Updated pom.xml to use newer versions of dependencies (httpclient, son, get-user, mockito, junit)
3. Added support for new in 3.12 On The Fly Filters API
MBU-8986, MBU-8975, MBU-8987, MBU-8988, MBU-8989
4. Added support for getting Tasks for a Group  API (Mambu issue fixed)
MBU-9015
5. Miscellaneous updates:
Removed deprecated in 3.11 methods
Add demo test support for User linked fields (MBU-8966) and for Preferred Language field in clients and groups (MBU-9221)

### CODE REVIEW

File: src/com/mambu/apisdk/services/ClientsService.java [hunk 1]
```diff
@@ -434,6 +435,27 @@ public GroupExpanded updateGroup(GroupExpanded groupDetails) throws MambuApiExce
...
 
 	/**
+	 * Get clients by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of clients matching filter constraint
+	 * @throws MambuApiException
+	 */
+	public List<Client> getClients(JSONFilterConstraints filterConstraints, String offset, String limit)
```
Comment (thobach) [inline src/com/mambu/apisdk/services/ClientsService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/ClientsService.java [hunk 2]
```diff
@@ -455,6 +477,27 @@ public GroupExpanded updateGroup(GroupExpanded groupDetails) throws MambuApiExce
...
 
 	/**
+	 * Get groups by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of groups matching filter constraint.
+	 * @throws MambuApiException
+	 */
+	public List<Group> getGroups(JSONFilterConstraints filterConstraints, String offset, String limit)
```
Comment (thobach) [inline src/com/mambu/apisdk/services/ClientsService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/LoansService.java [hunk 3]
```diff
@@ -548,6 +549,28 @@ public boolean patchLoanAccount(LoanAccount loan) throws MambuApiException {
...
 
+	/**
+	 * Get loan transactions by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of loan transactions matching filter constraint
+	 * @throws MambuApiException
+	 */
+	public List<LoanTransaction> getLoanTransactions(JSONFilterConstraints filterConstraints, String offset,
```
Comment (thobach) [inline src/com/mambu/apisdk/services/LoansService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/LoansService.java [hunk 4]
```diff
@@ -710,6 +733,27 @@ public LoanTransaction applyInterestToLoanAccount(String accountId, Date date, S
...
 
+	/**
+	 * Get loan accounts by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of loan accounts matching filter constraints
+	 * @throws MambuApiException
+	 */
+	public List<LoanAccount> getLoanAccounts(JSONFilterConstraints filterConstraints, String offset, String limit)
```
Comment (thobach) [inline src/com/mambu/apisdk/services/LoansService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/LoansService.java [hunk 5]
```diff
@@ -792,61 +836,4 @@ public LoanProduct getLoanProduct(String productId) throws MambuApiException {
...
-	 *             delete custom field values
-	 * 
-	 * @param accountId
-	 *            the encoded key or id of the Mambu Loan Account
-	 * @param customFieldId
-	 *            the encoded key or id of the custom field to be deleted
-	 * 
-	 * @throws MambuApiException
-	 */
-	public boolean deleteLoanAccountCustomField(String accountId, String customFieldId) throws MambuApiException {
```
Comment (thobach) [inline src/com/mambu/apisdk/services/LoansService.java]: Why did this method get deleted? It wasn't marked as deprecated.

File: src/com/mambu/apisdk/services/SavingsService.java [hunk 6]
```diff
@@ -254,6 +255,28 @@ public SavingsAccount undoApproveSavingsAccount(String accountId, String notes)
...
 
+	/**
+	 * Get savings transactions by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of savings transactions matching filter constraints
+	 * @throws MambuApiException
+	 */
+	public List<SavingsTransaction> getSavingsTransactions(JSONFilterConstraints filterConstraints, String offset,
```
Comment (thobach) [inline src/com/mambu/apisdk/services/SavingsService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/SavingsService.java [hunk 7]
```diff
@@ -254,6 +255,28 @@ public SavingsAccount undoApproveSavingsAccount(String accountId, String notes)
...
+			String limit) throws MambuApiException {
+		// Available since Mambu 3.12. See MBU-8988 for more details
+		// POST {JSONFilterConstraints} /api/savings/transactions/search?offset=0&limit=5
+
+		ApiDefinition apiDefintition = SearchService
+				.makeApiDefinitionforSearchByFilter(MambuEntityType.LOAN_TRANSACTION);
```
Comment (thobach) [inline src/com/mambu/apisdk/services/SavingsService.java]: This is confusing: SavingsTransaction vs. MambuEntityType.LOAN_TRANSACTION

File: src/com/mambu/apisdk/services/SavingsService.java [hunk 8]
```diff
@@ -618,6 +641,28 @@ public SavingsAccount closeSavingsAccount(String accountId, APIData.CLOSER_TYPE
...
 
+	/**
+	 * Get savings accounts by specifying filter constraints
+	 * 
+	 * @param filterConstraints
+	 *            filter constraints. Must not be null
+	 * @return list of savings accounts matching filter constraint
+	 * @throws MambuApiException
+	 */
+	public List<SavingsAccount> getSavingsAccounts(JSONFilterConstraints filterConstraints, String offset, String limit)
```
Comment (thobach) [inline src/com/mambu/apisdk/services/SavingsService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/SearchService.java [hunk 9]
```diff
@@ -94,4 +100,115 @@ public SearchService(MambuAPIService mambuAPIService) {
...
+	 * 
+	 * @param searchEntityType
+	 *            Mambu entity type. Must not be null. Currently searching with filter constrains API supports the
+	 *            following entities: Clients, Groups, Loans, Savings, Loan Transactions and SavingsTransactions.
+	 * @param filterConstraints
+	 *            filter constraints applicable for the entity. Must not be null
+	 * @return list of entities of the searchEntityType matching provided filter constraints
+	 * @throws MambuApiException
+	 */
+	public <T> List<T> searchEntities(MambuEntityType searchEntityType, JSONFilterConstraints filterConstraints,
```
Comment (thobach) [inline src/com/mambu/apisdk/services/SearchService.java]: Please add comment for params offset and limit.

File: src/com/mambu/apisdk/services/SearchService.java [hunk 10]
```diff
@@ -94,4 +100,115 @@ public SearchService(MambuAPIService mambuAPIService) {
...
+			String offset, String limit) throws MambuApiException {
+		// Available since Mambu 3.12. See MBU-8986 and MBU-8975 for more details
+		// Example: POST {JSONFilterConstraints} /api/savings/transactions/search?offset=0&limit=5
+
+		ApiDefinition apiDefintition = SearchService.makeApiDefinitionforSearchByFilter(searchEntityType);
```
Comment (thobach) [inline src/com/mambu/apisdk/services/SearchService.java]: Minor typo: apiDefintition

File: src/com/mambu/apisdk/services/TasksService.java [hunk 11]
```diff
@@ -154,6 +154,8 @@ public Task createTask(String title, String username, String description, Date d
 	 *            username, encodedKey or user id
 	 * @param clientId
 	 *            client encodedKey or id
```
Comment (thobach) [inline src/com/mambu/apisdk/services/TasksService.java]: I'd explain the "clientId != null && groupId != null" constraint here as well.

File: src/com/mambu/apisdk/util/RequestExecutorImpl.java [hunk 12]
```diff
@@ -114,20 +115,21 @@ public String executeRequest(String urlString, ParamsMap params, Method method,
 
 		}
 
+		HttpClient httpClient = new DefaultHttpClient();
```
Comment (thobach) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: I'd prefer using the "try with resources" approach used in https://hc.apache.org/httpcomponents-client-4.5.x/quickstart.html or the fluent API which both take good care of releasing resources.

File: src/com/mambu/apisdk/util/RequestExecutorImpl.java [hunk 13]
```diff
@@ -139,21 +141,22 @@ public String executeRequest(String urlString, ParamsMap params, Method method,
 		} catch (IOException e) {
 			LOGGER.warning("IOException: message= " + e.getMessage());
 			throw new MambuApiException(e);
+		} finally {
+			httpClient.getConnectionManager().shutdown();
```
Comment (thobach) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: Doesn't shutting down the connection manager prevent re-using the HTTP connection? Ideally we can re-use the connection for subsequent requests.

Comment (mdanilkis) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: I believe this only releases resources for this specific instance of the HttpClient (which it is about to go out of scope). I don't  think we have any code to ensure any connection persistence in the current implementation of the RequestExecutor in the wrapper.

File: src/com/mambu/apisdk/util/RequestExecutorImpl.java (hunk 12 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: For try with resources I think we would need to switch to CloseableHttpClient (from the HttpClient) but unfortunately the CloseableHttpClient is not supported by Android. So I left the HttpClient in place for now.  Can modify this when we decide how to proceed with wrapper migration.

File: src/com/mambu/apisdk/services/TasksService.java (hunk 11 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/TasksService.java]: OK, Will add to the params description as well.

File: src/com/mambu/apisdk/services/SearchService.java (hunk 9 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/SearchService.java]: OK. Will add pagination params comments in all places

File: src/com/mambu/apisdk/services/SavingsService.java (hunk 7 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/SavingsService.java]: Good catch. It's a bug and it needs to be changed to SAVINGS_TRANSACTION instead. Thanks!

File: src/com/mambu/apisdk/services/LoansService.java (hunk 5 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/services/LoansService.java]: This method was supposed to be marked as Deprecated. There was @deprecated for this method specified in the description but the "@Depracted" annotation was missing. Similar methods were have been now deleted for all other services (Clients, Branches, Savings, etc.) so this one should be deleted as well.

File: src/com/mambu/apisdk/util/RequestExecutorImpl.java (hunk 13 above)
Comment (thobach) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: I'd assume if we move the httpClient object to a field and reuse it everywhere, the lib takes care of connection persistence itself (http://hc.apache.org/httpclient-3.x/performance.html#Connection_persistence).

File: src/demo/DemoTestSearchService.java [hunk 14]
```diff
@@ -172,6 +194,149 @@ public static void testTypesCombinations() throws MambuApiException {
 
 	}
 
+	// Test API to GET entities by On The Fly Filter
+	private static void testSearchEntitiesByFilter() throws MambuApiException {
```
Comment (thobach) [inline src/demo/DemoTestSearchService.java]: Could you please add a few more test cases for:
1) multiple contraints in one filter
2) custom field filter
3) date filter using between to see if "second value" works

**Comment by thobach** [comment]
Very nice additions–well, done! Thanks Michael.

File: src/com/mambu/apisdk/util/RequestExecutorImpl.java (hunk 13 above)
Comment (mdanilkis) [inline src/com/mambu/apisdk/util/RequestExecutorImpl.java]: Should be. This should be done as a separate refactoring effort to do regression testing and to test for multithreading. potential leaks, etc.

File: src/demo/DemoTestSearchService.java (hunk 14 above)
Comment (mdanilkis) [inline src/demo/DemoTestSearchService.java]: Ok. Will add more filtering options.

**Pull request merged**